b2sdk/transfer/emerge/planner/planner.py

```diff
@@ -0,0 +1,493 @@
+#
+# License https://www.backblaze.com/using_b2_code.html
+#
+######################################################################
+
+import hashlib
+import json
+
+from abc import ABCMeta, abstractmethod
+from collections import deque
+from itertools import chain
+
+import six
+
+from b2sdk.transfer.emerge.planner.part_definition import (
+    CopyEmergePartDefinition,
+    UploadEmergePartDefinition,
+    UploadSubpartsEmergePartDefinition,
+)
+from b2sdk.transfer.emerge.planner.upload_subpart import (
+    LocalSourceUploadSubpart,
+    RemoteSourceUploadSubpart,
+)
+
+MEGABYTE = 1000 * 1000
+GIGABYTE = 1000 * MEGABYTE
+
+
+class EmergePlanner(object):
+    def __init__(
+        self,
+        min_part_size=5 * MEGABYTE,
+        recommended_upload_part_size=100 * MEGABYTE,
+        max_part_size=5 * GIGABYTE
+    ):
+        self.min_part_size = min(min_part_size, recommended_upload_part_size, max_part_size)
+        self.recommended_upload_part_size = min(recommended_upload_part_size, max_part_size)
+        self.max_part_size = max_part_size
+
+    @classmethod
+    def from_account_info(
+        cls,
+        account_info,
+        min_part_size=None,
+        recommended_upload_part_size=None,
+        max_part_size=None
+    ):
+        # TODO: add support for getting `min_part_size` and `max_part_size` from account info
+        if recommended_upload_part_size is None:
+            # TODO: change `get_minimum_part_size` to correct name
+            recommended_upload_part_size = account_info.get_minimum_part_size()
+        kwargs = {
+            'min_part_size': min_part_size,
+            'recommended_upload_part_size': recommended_upload_part_size,
+            'max_part_size': max_part_size,
+        }
+        return cls(**{key: value for key, value in kwargs.items() if value is not None})
+
+    def get_emerge_plan(self, write_intents):
+        write_intents = sorted(write_intents, key=lambda intent: intent.destination_offset)
+        return self._get_emerge_plan(write_intents, EmergePlan)
+
+    def get_streaming_emerge_plan(self, write_intent_iterator):
+        return self._get_emerge_plan(write_intent_iterator, StreamingEmergePlan)
+
+    def _get_emerge_plan(self, write_intent_iterator, plan_class):
+        return plan_class(
+            self._get_emerge_parts(
+                self._select_intent_fragments(self._validatation_iterator(write_intent_iterator))
+            )
+        )
+
+    def _get_emerge_parts(self, intent_fragments_iterator):
+        min_part_size = self.min_part_size
+        buff_start = 0
+        buff_end = 0
+        buff = []
+        current_end = 0
+        current_intent = None
+        for intent, fragment_end in intent_fragments_iterator:
+            if current_intent is None:
+                current_intent = intent
+                current_end = fragment_end
+                continue
+
+            if intent is current_intent:
+                current_end = fragment_end
```

Comment:
also, why? if we have duplicate intents on the input iterator, is it going to be skipped? why?

Reply:
intents cannot be duplicated - if you put the same object (`WriteIntent`) on interator, "chooser" part would not care about which to choose, and it is actually not important, because it does not matter - even if you put "same" write intents (meaning with the same outbound source and destination offset) - chooser would choose only one - BUT - here it just merging two fragments of the same write intent into one